doc/src/manual/calling-c-and-fortran-code.md

```diff
@@ -854,18 +854,135 @@ it must be handled in other ways.
+
+A `LazyLibrary` represents a library that opens itself (and its dependencies) automatically
+on first use in a `ccall()`, `@ccall`, `dlopen()`, `dlsym()`, `dlpath()`, or `cglobal()`.
+The library is loaded exactly once in a thread-safe manner, and subsequent calls reuse the
+loaded library handle.
+
+#### Basic Usage
+
+```julia
+using Libdl
+
+# Define a LazyLibrary as a const for optimal performance
+const libz = LazyLibrary("libz")
+
+# Use directly in @ccall - library loads automatically on first call
+@ccall libz.deflate(strm::Ptr{Cvoid}, flush::Cint)::Cint
+
+# Also works with ccall
+ccall((:inflate, libz), Cint, (Ptr{Cvoid}, Cint), strm, flush)
+```
+
+#### Platform-Specific Libraries
+
+For code that needs to work across different platforms:
+
+```julia
+const mylib = LazyLibrary(
+    if Sys.iswindows()
+        "mylib.dll"
+    elseif Sys.isapple()
+        "libmylib.dylib"
+    else
+        "libmylib.so"
+    end
+)
+```
+
+#### Libraries with Dependencies
+
+When a library depends on other libraries, specify the dependencies to ensure
+they load in the correct order:
+
+```julia
+const libfoo = LazyLibrary("libfoo")
+const libbar = LazyLibrary("libbar"; dependencies=[libfoo])
+
+# When libbar is first used, libfoo is loaded first automatically
+@ccall libbar.bar_function(x::Cint)::Cint
+```
+
+#### Lazy Path Construction
+
+For libraries whose paths are determined at runtime, use `LazyLibraryPath`:
+
+```julia
+# Path is constructed when library is first accessed
+const mylib = LazyLibrary(LazyLibraryPath(artifact_dir, "lib", "libmylib.so"))
+```
+
+#### Initialization Callbacks
+
+If a library requires initialization after loading:
+
+```julia
+const mylib = LazyLibrary("libmylib";
+    on_load_callback = () -> @ccall mylib.initialize()::Cvoid
+)
+```
+
+!!! warning
+    The `on_load_callback` should be minimal and must not call `wait()` on any tasks.
+    It is called exactly once by the thread that loads the library.
+
+#### Conversion from `__init__()` Pattern
+
+Before `LazyLibrary`, library paths were often computed in `__init__()` functions.
+This pattern can be replaced with `LazyLibrary` for better performance and thread safety.
+
+Old pattern using `__init__()`:
+
+```julia
+# Old: Library path computed in __init__()
+libmylib_path = ""
+
+function __init__(
+    # Loads library on startup, whether it is used or not
+    global libmylib_path = find_library(["libmylib"])
+end
+
+function myfunc(x)
+    ccall((:cfunc, libmylib_path), Cint, (Cint,), x)
+end
+```
+
+New pattern using `LazyLibrary`:
+
+```julia
+# New: Library as const, no __init__() needed
+const libmylib = LazyLibrary("libmylib")
+
+function myfunc(x)
+    # Library loads automatically just before calling `cfunc`
+    @ccall libmylib.cfunc(x::Cint)::Cint
+end
+```
+
+For more details, see the [`Libdl.LazyLibrary`](@ref) documentation.
+
+### Overloading `dlopen` for Custom Types
```

Comment:
What is this section for?  Are you trying to warn other people who might overload `dlopen()` (like we did for `LazyLibrary`) or are you saying there's some reason someone might want to overload `dlopen()` for `LazyLibrary` itself?

Reply:
You can write your own LazyLibrary-like types. This is documentation for those fools who write LazyLibrary types, like us.